daft/datatype.py

```diff
@@ -119,135 +116,214 @@ def _infer_type(cls, user_provided_type: DataTypeLike) -> DataType:
+        elif dt.is_struct():
+            fields = dt.get_struct_inner()
+            return StructType({name: DataType._from_pydatatype(datatype) for name, datatype in fields.items()})
+        elif dt.is_map():
+            (key, value) = dt.get_map_inner()
+            return MapType(DataType._from_pydatatype(key), DataType._from_pydatatype(value))
+        elif dt.is_tensor():
+            inner = dt.get_tensor_inner()
+            return TensorType(DataType._from_pydatatype(inner))
+        elif dt.is_fixed_shape_tensor():
+            (inner, shape) = dt.get_fixed_shape_tensor_inner()
+            print(f"inner: {inner}, shape: {shape}")
+            return FixedShapeTensorType(dtype=DataType._from_pydatatype(inner), shape=shape)
+        elif dt.is_sparse_tensor():
+            (inner, use_offset_indices) = dt.get_sparse_tensor_inner()
+            return SparseTensorType(DataType._from_pydatatype(inner), use_offset_indices)
+        elif dt.is_fixed_shape_sparse_tensor():
+            (inner, shape, use_offset_indices) = dt.get_fixed_shape_sparse_tensor_inner()
+            return FixedShapeSparseTensorType(DataType._from_pydatatype(inner), shape, use_offset_indices)
+        elif dt.is_image():
+            mode = dt.get_image_inner()
+            return ImageType(mode)
+        elif dt.is_fixed_shape_image():
+            (mode, height, width) = dt.get_fixed_shape_image_inner()
+            return FixedShapeImageType(mode, height, width)
+        elif dt.is_embedding():
+            (inner, size) = dt.get_embedding_inner()
+            return EmbeddingType(DataType._from_pydatatype(inner), size)
+        elif dt.is_decimal():
+            (precision, scale) = dt.get_decimal_inner()
+            return DecimalType(precision, scale)
+        elif dt.is_extension():
+            (name, storage_dtype, metadata) = dt.get_extension_inner()
+            return ExtensionType(name, DataType._from_pydatatype(storage_dtype), metadata)
+        raise ValueError(f"Unhandled PyDataType: {dt}. This conversion should be exhaustive.")
 
     @classmethod
-    def int8(cls) -> DataType:
+    def int8(cls) -> Int8Type:
         """Create an 8-bit integer DataType."""
-        return cls._from_pydatatype(PyDataType.int8())
+        return Int8Type()
 
     @classmethod
-    def int16(cls) -> DataType:
+    def int16(cls) -> Int16Type:
         """Create an 16-bit integer DataType."""
-        return cls._from_pydatatype(PyDataType.int16())
+        return Int16Type()
 
     @classmethod
-    def int32(cls) -> DataType:
+    def int32(cls) -> Int32Type:
         """Create an 32-bit integer DataType."""
-        return cls._from_pydatatype(PyDataType.int32())
+        return Int32Type()
 
     @classmethod
-    def int64(cls) -> DataType:
+    def int64(cls) -> Int64Type:
         """Create an 64-bit integer DataType."""
-        return cls._from_pydatatype(PyDataType.int64())
+        return Int64Type()
```

Comment:
Upon deeper review, I actually think the pattern that engines like pyarrow follow is to leverage singletons for many of the types [primitives and unparameterized ones]. The Classes are simply an abstraction to the user.

See [function](https://github.com/apache/arrow/blob/9b0885ade1c8bbefa173e79391c2f1a61b93456d/python/pyarrow/types.pxi#L4000) to create type: 

and where it actually get [constructed](https://github.com/apache/arrow/blob/9b0885ade1c8bbefa173e79391c2f1a61b93456d/python/pyarrow/types.pxi#L3816):